src/System.Buffers.Experimental/System/Buffers/Text/BytesReader.cs

```diff
@@ -109,81 +93,129 @@ public PositionRange ReadRange(ReadOnlySpan<byte> delimiter)
+                    Debug.Assert(count == currentSegmentUnread);
+                    _currentSpan = Span<byte>.Empty;
+                    _currentSpanIndex = 0;
+                    return;
                 }
-                _currentSpan = _currentMemory.Span;
-                _currentSpanIndex = 0;
-                Seek(count - unreadLength);
             }
+            _currentSpan = memory.Span;
+            _currentSpanIndex = 0;
+            Advance(count - currentSegmentUnread);
         }
 
-        public void Seek(Position position)
+        public void Advance(Position position)
         {
             _currentSegmentPosition = position;
             _nextSegmentPosition = position;
-            if (_bytes.TryGet(ref _nextSegmentPosition, out _currentMemory))
+            if (_bytes.TryGet(ref _nextSegmentPosition, out ReadOnlyMemory<byte> memory))
             {
-                _currentSpan = _currentMemory.Span;
+                _currentSpan = memory.Span;
             }
             else
             {
                 _currentSpan = default;
             }
             _currentSpanIndex = 0;
-
         }
 
         #region Parsing Methods
         // TODO: how to we chose the lengths of the temp buffers?
         // TODO: these methods call the slow overloads of Parsers.Custom.TryParseXXX. Do we need fast path?
         // TODO: these methods hardcode the format. Do we need this to be something that can be specified?
-        public bool TryParseBoolean(out bool value)
+        public bool TryParse(out bool value)
         {
             var unread = Unread;
-            if (CustomParser.TryParseBoolean(unread, out value, out int consumed, _symbolTable))
+            if (Utf8Parser.TryParse(unread, out value, out int consumed))
             {
-                Debug.Assert(consumed <= unread.Length);
                 if (unread.Length > consumed)
                 {
                     _currentSpanIndex += consumed;
                     return true;
                 }
             }
 
-            Span<byte> tempSpan = stackalloc byte[15];
-            CopyTo(this, tempSpan, out var copied);
+            Span<byte> tempSpan = stackalloc byte[5];
+            var copied = CopyTo(this, tempSpan);
 
-            if (CustomParser.TryParseBoolean(tempSpan.Slice(0, copied), out value, out consumed, _symbolTable))
+            if (Utf8Parser.TryParse(tempSpan.Slice(0, copied), out value, out consumed))
             {
-                Seek(consumed);
+                Advance(consumed);
                 return true;
             }
 
             return false;
         }
 
-        public bool TryParseUInt64(out ulong value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryParse(out int value)
+        {
+            var unread = Unread;
+            if (Utf8Parser.TryParse(unread, out value, out int consumed))
+            {
+                if (unread.Length > consumed) {
```

Comment:
Should this be `if (unread.Length >= consumed)`?

Reply:
No, the check is here to ensure that the parser does not "quit" early when there are more digits in the subsequent segment, i.e. [1, 2, 3] [4, x] should parse as 1234, not 123.